File: flyappy_autonomy_code_py/src/flyappy_autonomy_code/flyappy_ros.py
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Vector3
from std_msgs.msg import Bool


class FlyappyRos:
    def __init__(self, node: Node):
        self._logger = node.get_logger()

        # Publisher for sending acceleration commands to Flyappy
        self._pub_acc_cmd = node.create_publisher(
            Vector3,
            "/flyappy_acc",
            1
        )

        # Subscribers to topics from Flyappy game
        self._sub_vel = node.create_subscription(
            Vector3,
            "/flyappy_vel",
            self.velocity_callback,
            10
        )
        self._sub_laser_scan = node.create_subscription(
            LaserScan,
            "/flyappy_laser_scan",
            self.laser_scan_callback,
            10
        )
        self._sub_game_ended = node.create_subscription(
            Bool,
            "/flyappy_game_ended",
            self.game_ended_callback,
            5
        )

    def velocity_callback(self, msg: Vector3) -> None:
        # Example of publishing acceleration command to Flyappy
        x = 0.5  # move and accelerate Flyappy forward
        y = 0
        self._pub_acc_cmd.publish(Vector3(x=float(x), y=float(y), z=float(0)))

    def laser_scan_callback(self, msg: LaserScan) -> None:
        # Example of printing laser angle and range
        self._logger.info(
            f"Laser range: {msg.ranges[0]}, angle: {msg.angle_min}", throttle_duration_sec=1

        )

    def game_ended_callback(self, msg: Bool) -> None:
        if msg.data:
            self._logger.info("Crash detected.")
        else:
            self._logger.info("End of countdown.")

File: flyappy_autonomy_code_cpp/src/flyappy/flyappy_ros.cpp
#include "flyappy/flyappy_ros.hpp"

namespace flyappy
{

inline constexpr uint32_t QUEUE_SIZE = 5u;

FlyappyRos::FlyappyRos(rclcpp::Node::SharedPtr node) : node_(node)
{
    pub_acceleration_command_ = node_->create_publisher<geometry_msgs::msg::Vector3>(
            "/flyappy_acc", QUEUE_SIZE);
    sub_velocity_ = node_->create_subscription<geometry_msgs::msg::Vector3>(
            "/flyappy_vel", QUEUE_SIZE,
            std::bind(&FlyappyRos::velocityCallback, this, std::placeholders::_1));
    sub_laser_scan_ = node_->create_subscription<sensor_msgs::msg::LaserScan>(
            "/flyappy_laser_scan", QUEUE_SIZE,
            std::bind(&FlyappyRos::laserScanCallback, this, std::placeholders::_1));
    sub_game_ended_ = node_->create_subscription<std_msgs::msg::Bool>(
            "/flyappy_game_ended", QUEUE_SIZE,
            std::bind(&FlyappyRos::gameEndedCallback, this, std::placeholders::_1));
}

void FlyappyRos::velocityCallback([[maybe_unused]] const geometry_msgs::msg::Vector3& msg)
{
    // Example of publishing acceleration command to Flyappy
    geometry_msgs::msg::Vector3 acc_cmd;
    acc_cmd.x = 0.5;  // move and accelerate Flyappy forward
    acc_cmd.y = 0;
    pub_acceleration_command_->publish(acc_cmd);
}

void FlyappyRos::laserScanCallback(const sensor_msgs::msg::LaserScan& msg)
{
    // Example of printing laser angle and range
    RCLCPP_INFO_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
                         "Laser range: %f, angle: %f", msg.ranges[0], msg.angle_min);
}

void FlyappyRos::gameEndedCallback(const std_msgs::msg::Bool& msg)
{
    if (msg.data)
    {
        RCLCPP_INFO(node_->get_logger(), "Crash detected.");
    }
    else
    {
        RCLCPP_INFO(node_->get_logger(), "End of countdown.");
    }

    flyappy_ = Flyappy{};
}

}  // namespace flyappy
